### PR TITLE
[Bug fix] Llama Galaxy sampling: per-request seeding, prefill logit routing, and mixed greedy/random decode

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
Model-level sampling fixes for Llama-Galaxy: deterministic per-request RNG seeding that survives batch operations, prefill last-token logit routing, mixed greedy/random "doubling", penalty accounting, and removal of debug-logging scaffolding.

## Changes & rationale
1. Prefill last-token logit routing (llama_model.py, generator.py)
Per-user logits were previously saved once after the split loop, capturing only the last iteration's tt_logits. In a batched prefill, every other user's row in the host buffer stayed zero-initialized, and when logprobs are requested the batch is host-sampled from those rows, so affected users got a garbage token (argmax over a uniform vector) and -log(vocab) logprobs. Fix: save each user's logits inside the loop (tt_out_logits_saved has one row per split entry; row i ↔ x_split[i]). The return-logits/PCC path likewise now fills one row per slot for batched prefill (was always 1 row, dropping all but the last user).

2. Mixed greedy/random "doubling" (generator.py)
Mixed greedy+random batches use device-resident tokens/positions (like all-greedy): the sampled token is fed back on-device and current_pos advances in-trace. Forcing a per-step host reload routes the device-sampled token through a host readback that, under async/traced overlap, is enqueue-mis-ordered vs the next step's trace overwrite → interleaved stale/current tokens → deterministic garbage. The residual reset-driven greedy "doubling" is fixed by preserving device current_pos across page-table / reset_batch reloads: under on-device sampling a page-table change refreshes only the page-table trace input (not a full reload), and the transition-step drain lives in the lane runner.

3. Deterministic per-request seeding (generator.py, tt_penalties.py)
Seeds survive batch condense and slot remap. Explicit request seeds derive the per-token device seed from the absolute decode position, so a stream stays reproducible even when vLLM temporarily removes a running request and re-admits it in a different slot. apply_slot_remap moves a condensed slot's RNG state (seed, counter, generator) into its new slot and clears the vacated source, matching what vLLM's InputBatch.condense actually emits. Penalty token accounting corrected.

4. Debug-logging scaffolding removed (_utils.py, generator.py, tt_sampling.py).
Temporary (tt-metal-side) - tracked for vLLM
The reload decisions (item 2) are a per-model workaround: only vLLM knows when host inputs are valid under async draining, so today every model has to re-derive that logic. Marked TEMP FIX (tenstorrent/vllm#449) in code and tracked in tenstorrent/vllm#449, to be moved into vLLM (incl. a "reload only page table" signal) and removed here, then retested.

## Tests
models/common/tests/test_sampling.py covers the slot-remap contract host-only: a condense relabels the destination slot and vacates the source, and the identity steady state is a no-op. Serving-level seeding behaviour is already covered by the vLLM plugin suite (plugins/vllm-tt-plugin/tests/tt/test_seeding_and_variety.py) and is not duplicated here.

## Impact
Reproducible seeded generation and deterministic greedy decode; no
doubled/duplicated tokens on mixed batches.

Note: the device-side counterpart to the doubling fix lives in the vLLM TT plugin
(wait_for_all_pending_async_steps() drain on layout change) and is already on vLLM dev.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:model-sampling-seeding)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:model-sampling-seeding)

### CODE REVIEW

**Review by mtairum — APPROVED** [review]
Approving to unblock. @ifabijanic-tt are we running these new tests anywhere?

Should we?

**Comment by ifabijanic-tt** [comment]
Tests will be added to CI right after PR is merged

File: models/common/sampling/generator.py [hunk 1]
```diff
@@ -634,6 +601,9 @@ def __init__(self, tt_sampling, max_batch_size=32):
         # True only for the most recent get_new_values() call when at least
         # one active slot used an explicit request seed.
         self._active_request_seed = False
+        # Last non-identity slot remap applied, used to ignore a steady-state
+        # remap that is (incorrectly) re-emitted verbatim every decode step.
+        self._last_applied_remap = None
```
Comment (tchedaTT) [inline models/common/sampling/generator.py]: When is a remap reemitted? We should fix this on the vllm side rather than adding a workaround to each model

File: models/common/sampling/generator.py [hunk 2]
```diff
@@ -782,42 +739,42 @@ def apply_slot_remap(self, remap):
...
             self.rngs[new_slot] = copy.copy(old_rngs[old_slot])
-        for old_slot in moved_sources - moved_destinations:
-            self.seeds[old_slot] = None
-            self.seed_counters[old_slot] = 0
+        # NOTE: deliberately do not clear "moved source" slots to None here.
+        # A source that was genuinely vacated by a condense is harmless to leave
+        # populated (an inactive slot is pushed MAX_UINT32/SKIP and is never read
+        # back, and a reused slot is reset via reset_seed). Clearing was unsafe
+        # because a replication map lists real, still-active requests as sources;
+        # zeroing them dropped reproducible seeds and broke determinism.
```
Comment (tchedaTT) [inline models/common/sampling/generator.py]: I don't get why we can't clear slots we moved from? This doesn't sound correct

File: models/demos/llama3_70b_galaxy/tt/llama_model.py [hunk 3]
```diff
@@ -614,11 +614,16 @@ def process_output_prefill(self, tt_out, last_token_idx, tt_out_logits_saved=Non
...
-            logits_saved = ttnn.to_torch(ttnn.get_device_tensors(tt_logits)[output_device_idx]).float()[0, 0, :, :]
-
-            tt_out_logits_saved.copy_(logits_saved)
+            # Save each user's logits INSIDE the loop. Previously this was done once
+            # after the loop and captured only the last iteration's tt_logits, so in a
+            # batched prefill every other user's row in the caller's host buffer stayed
+            # zero-initialized. When logprobs are requested the batch is host-sampled
+            # from those rows, so the affected users got a garbage token (argmax over a
+            # uniform vector) and -log(vocab) logprobs. tt_out_logits_saved has one row
+            # per split entry; row i corresponds to x_split[i] (slot i for batched).
```
Comment (tchedaTT) [inline models/demos/llama3_70b_galaxy/tt/llama_model.py]: Rationale for change / debug explanation should go in PR description, not here

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 4]
```diff
@@ -739,10 +630,14 @@ def prefill_forward_text(
             if not work_use_batched_prefill:
                 prefill_kwargs["num_cached_tokens"] = num_cached_tokens_list[request_idx]
 
-            # Save output logits (PCC check / return_logits path)
+            # Save output logits (PCC check / return_logits path). Batched prefill fills
+            # one row per slot (padded_batch rows); non-batched fills a single row.
+            # (Previously always 1 row, which silently dropped all but the last batched
+            # user's logits -> garbage host-sampled tokens/logprobs for the rest.)
```
Comment (tchedaTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py]: Unnecessary comment about past state

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 5]
```diff
@@ -774,7 +669,12 @@ def prefill_forward_text(
                     output_toks[request_idx] = tt_tok
 
                 if tt_out_logits_all_users is not None and tt_out_logits_saved is not None:
```
Comment (tchedaTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py:669]: Do we still use this returning path anywhere, or is this dead code?

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 6]
```diff
@@ -1402,35 +1276,37 @@ def decode_forward(
...
+        # Mixed greedy+random batches intentionally do NOT force a host input reload here:
+        # like all-greedy batches they use device-resident tokens/positions (sampled token
+        # fed back on-device, current_pos advanced in-trace). Forcing a per-step host reload
+        # would route the device-sampled token through a host readback that, under async/traced
+        # overlap, is enqueue-mis-ordered vs the next step's trace overwrite -> interleaved
+        # stale/current tokens -> deterministic garbage for the whole mixed batch. Per-request
+        # sampling params are fixed for a request's lifetime (uploaded at batch setup / refreshed
+        # on reset_batch), so no per-step re-upload is needed. The residual reset-driven greedy
+        # "doubling" is handled by preserving device current_pos across page_table / reset_batch
+        # reloads -- see the page_table_changed branch below and the pre-reload drain in the lane runner.
```
Comment (tchedaTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py]: Unnecessary comment about past state - should go in PR description

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 7]
```diff
@@ -1402,35 +1276,37 @@ def decode_forward(
             reset_inputs = True
             reset_reasons.append("decode_inputs_need_reset")
             self._decode_inputs_need_reset = False
```
Comment (tchedaTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py:1276]: What about this and all other cases where the model decides to reload, even though vllm didn't set reset_batch? How do we know the inputs aren't stale then, like they were in the mixed greedy/random case?
I think the solution is to make all reloading decisions in vllm, because only it knows when inputs are valid under async - the logic has to match the async draining logic. It would also have the advantage of not having to replicate in every model.

Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py:1276]: that would entail adding another arg for "reload only page table"

**Review by tchedaTT — CHANGES_REQUESTED** [review]

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 7 above)
Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py:1276]: Consensus: There are probably still more hidden issues that need to be resolved comprehensively. Created https://github.com/tenstorrent/vllm/issues/449 to track the wider issue

File: models/common/sampling/generator.py (hunk 1 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: Remap is re-emitted every decode step: vLLM hands the same non-identity replication/layout, not a one-shot condense delta. We can keep the model-side deduplication as TEMP FIX but we will need to move it move it to vLLM.

File: models/common/sampling/generator.py (hunk 2 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: The remap can be a replication map where the source slot is a real, still-active request and clearing it to None drops that request's reproducible seed and breaks determinism. Genuinely vacated sources are harmless (inactive -> pushed SKIP, reused -> reset_seed).

File: models/demos/llama3_70b_galaxy/tt/llama_model.py (hunk 3 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/llama_model.py]: Agree! Done, its trimmed to current-state note.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 4 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Also removed

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 5 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py:669]: It's the return-logits/PCC-check path, used by demo/PCC callers that pass a logits buffer back, not the vLLM serving path itself. Keep (or gate behind the logits flag). But the reset_seed_from_slots_if_needed's path is genuinely unused by both callers so its safe to drop the return.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 6 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Makes sense, removed.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 7 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py:1276]: Agree, because of that all current fixes on tt-metal side are marked with TEMP FIX (PR #50685) so when this issue is resolved the tt-metal side will be reverted.

File: models/common/sampling/generator.py (hunk 1 above)
Comment (tchedaTT) [reply models/common/sampling/generator.py]: I really think this one should go directly into vllm now, with no temp fixes, if we actually re-emit the remap multiple times instead of once.

File: models/common/sampling/generator.py (hunk 2 above)
Comment (tchedaTT) [reply models/common/sampling/generator.py]: How does it break determinism if we already moved the request to the new location?

File: models/demos/llama3_70b_galaxy/tt/llama_model.py (hunk 3 above)
Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/llama_model.py]: I don't see the change?

Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/llama_model.py]: Now its here :)

File: models/common/sampling/generator.py (hunk 2 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: Because the remap doesn't tell us the request left the source slot, it's a gather, not a permutation. remap[i] = j means "slot i now reads the state that was at slot j"; nothing in the map says slot j is now empty. Vacancy is then signalled separately (start_pos < 0), so "moved from" can't be inferred from the map.

The case that actually gives us issues on Galaxy is that vLLM replicates rank 0's requests across DP ranks and re-emits the same map every decode step. For example, remap = [0, 1, 0, 1] — slots 2,3 mirror slots 0,1, and 0,1 are still the live, running requests. Without this change that gives moved_sources: moved_destinations = {0, 1}, so both live requests got seeds[slot] = None. With that a seeded request silently becomes unseeded, falls back to the per-slot boot RNG, and stops being reproducible run-to-run. Even worse, once enough slots are cleared, __seed_active_ flips to False and _apply_slot_remap_ early-returns from then on, so subsequent remaps aren't applied at all.

File: models/common/sampling/generator.py (hunk 1 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: So, should I remove it from this commit then? If I have a guarantee that this will be in vLLM we can do that.

Comment (tchedaTT) [reply models/common/sampling/generator.py]: I am not convinced vllm really emits a non-identity remap multiple times for one layout change. I couldn't find such scenario and neither could an agent when asked.
If it really does, this needs to be fixed in vllm, please file an issue with a repro/explanation or a PR.
If it doesn't, this workaround is a red herring and shouldn't be here.
My intuition is this might be a false lead for sth really caused by the bug https://github.com/tenstorrent/vllm/pull/444/ is fixing.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 4 above)
Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: I don't think the removal was pushed

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 6 above)
Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: also not pushed

File: models/common/sampling/generator.py (hunk 2 above)
Comment (tchedaTT) [reply models/common/sampling/generator.py]: I don't think [0,1,0,1] is a valid value of slot_remap. We don't allow replicating requests, and in the gathered dp case the slot numbers are mapped to global first.

File: models/common/sampling/generator.py (hunk 1 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: You're actually right, and it could be the artifact of working on previous vllm plugin version, current version as much as I (and agent) can see don't have this issueI and I also can't see the scenario either. As much as I can see the whole remap thing was observed in mid-June on the multi-process DP Galaxy path (world=4, B=8), couple of days before Galaxy moved to lane-DP, where condense doesn't exist and the remap is structurally always identity.

Because of that it's removed now.

File: models/common/sampling/generator.py (hunk 2 above)
Comment (ifabijanic-tt) [reply models/common/sampling/generator.py]: Actually yeah, you are correct, the [0,1,0,1] isn't reachable. The clearing is restored back.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 4 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Yeah, for some reason the removal wasn't pushed... Fixed now.

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 6 above)
Comment (ifabijanic-tt) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Yeah, for some reason the removal wasn't pushed... Fixed now.

**Comment by ifabijanic-tt** [comment]
All changes from comments are added.

File: tools/vllm/test_non_uniform_seeding.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent USA, Inc.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tchedaTT) [inline tools/vllm/test_non_uniform_seeding.py]: This seems to duplicate the sampling test in https://github.com/tenstorrent/vllm/tree/dev/plugins/vllm-tt-plugin/tests/tt

**Review by tchedaTT — APPROVED** [review]
Left one last comment. Please update PR description and title to reflect new scope and re-test (CI is super slow now so locally is ok by me), then ok to merge

File: tools/vllm/test_non_uniform_seeding.py (hunk 8 above)
Comment (ifabijanic-tt) [reply tools/vllm/test_non_uniform_seeding.py]: Yeah, agree, and the sampling test in vllm is more detailed as well. Removed.

**Comment by ifabijanic-tt** [comment]
Local testing was done on wh-glx6u-04 machine. Targeted custom decode input reset policy test was run 20x all passing green. Reverting fixes in this PR reintroduces bugs so this is a good signal.
Full Llama 70B test was also ran 5x on vLLM and while we still have token doubling issues (issues tackled in other PRs and vLLM PR), all related miixed greedy/random doubling issues are not a problem anymore. And per request seeding is now deterministic.
The PR will be merged.
